python/dgl/graphbolt/negative_sampler.py

```diff
@@ -0,0 +1,132 @@
+"""Negative samplers"""
+
+
+from collections.abc import Mapping
+
+import torch
+
+from .graph_storage import CSCSamplingGraph
+from .linked_data_format import LinkedDataFormat
+
+__all__ = ["PerSourceUniformSampler"]
+
+
+class _BaseNegativeSampler:
+    """
+    A negative sampler used to generate negative samples and return
+    a mix of positive and negative samples, the format of the output
+    depends on the specified `linked_data_format`.
+    """
+
+    def __init__(
+        self,
+        graph: CSCSamplingGraph,
+        negative_ratio: int,
+        linked_data_format: LinkedDataFormat,
```

Comment:
For neighbor sampling we want to support 2 format since users may prepare the negative edges in training set directly. For our negative sampling canonical implementation, likely it is not necessary to output 2 formats, we can just pick one format that easier to be used in the neighbor sampling.